### PR TITLE
kubelet --node-ip flag using  unspecified IPs and external cloud provider node addresses behavior

### LINKED ISSUE #125348: Change in `kubelet --node-ip` behavior in 1.29 breaks some deployment models on clouds
(from #125300)

#121028 changed kubelet's startup behavior when using an external cloud provider, so that it no longer tries to guess the right primary node IP before the cloud-controller-manager starts (since in the case where it guessed wrong, this would cause problems). But this means it can no longer set `pod.status.hostIP` / host-network `pod.status.podIP` on pods created before the ccm starts, which breaks some deployment models involving using static pods to bootstrap the cluster. (In particular, this is breaking some OpenShift functionality, which assumed that kubelet could deploy etcd and/or kube-apiserver via static pods and pass the node IP to them via the downward API.)

One suggested workaround for this problem is for the administrator to pass `--node-ip` to kubelet, but that suggestion runs into the problem that doing this would cause the node's `node.status.addresses` to be stripped of secondary addresses, which may have further side effects, if there is software that assumes that all node IPs will be listed in `node.status.address`. #125300 considered changing the way ccm treats `--node-ip` to make it _not_ strip secondary node IPs when `--node-ip` was used, but as pointed out there, this _also_ threatens to break other things, since some existing software may be depending on the fact that this behavior _does_ happen.

We briefly thought that the problem could be solved by passing `--node-ip 0.0.0.0`, but it turns out that this doesn't actually do anything different from just not passing `--node-ip`.

/sig network
/sig node

### PR DESCRIPTION
**Pull request by aojea**
/kind documentation
/kind bug

When the kubelet is using a external cloud provider and starts with the -node-ip flag set to an IP address (except 0.0.0.0 or ::), it creates the Node object with that IP and adds an annotation `alpha.kubernetes.io/provided-node-ip` to the Node  to indicate the cloud provider to use only that IP on the node.status.addresses. In theory, this filtering behavior can not be changed because it can break the node approver CSR logic https://github.com/kubernetes/kubernetes/pull/125300#issuecomment-2148369463 , although is debatable the CSR logic is completely correct.

Before https://github.com/kubernetes/kubernetes/pull/121028 , the kubelet , if no --node-ip was set, started populating the node.status.addresses based on default discovery mechanism , and later the external cloud provider override them. Since the default discovery is "use the IP associated to the default gateway for IPv4", this breaks IPv6 clusters where all the hostNetwork Pods created before the cloud controller manager could update the addresses, will have IPv4 addresses.

The main problem now is to support external cloud providers that are deployed as Pods, since this present a chicken an egg problem, that the external cloud provider will need an initial valid IP to be populated. Using the node-ip flag with one IP can solve the problem but has some edges, it requires to know the IP beforehand and it filter the rest of IPs the cloud provider may want to add and other components will depend on them.

The node-ip flag already has different semantics for the unspecified addresses `0.0.0.0` or `::` to mean "discovery", as in _use the IP associated to the default gateway of the corresponding IP family_. This PR gives the same meaning to this flag when used with external cloud providers, and it allows to bootstrap the node with the IP of the desired IP family, and it allows to override it by the external cloud provider without filtering.

```release-note
The kubelet, when using --cloud-provider=external can use the --node-ip flag with one of the unspecified addresses 0.0.0.0 or ::, to create the Node with the IP of the default gateway of the corresponding IP family and then delegating the responsibility to the external cloud provider. This solve the bootstrap problems of out of tree cloud providers that are deployed as Pods within the cluster.
```

Fixes: #125348

### CODE REVIEW

**Review by danwinship — COMMENTED** [review]
needs a new unit test in `pkg/kubelet/setters_test.go` to cover the newly-documented behavior, but I think it would be easier for me to just write it than to describe what I want it to do...

**Comment by danwinship** [comment]
> needs a new unit test in `pkg/kubelet/setters_test.go` to cover the newly-documented behavior, but I think it would be easier for me to just write it than to describe what I want it to do...

hm, no, it needs updates to `NodeAddresses` for the autodetection to be mockable. And of course, NodeAddresses should probably be simplified to remove the legacy cloud provider case at this point.

Anyway, what we need is a test that, for various values of `--node-ip` including "not set" and "not the same IP family the cloud expects", tests what the first call to `NodeAddresses()` would do (as with the existing unit test), then overwrites `node.status.addresses` (simulating the update from the cloud provider), then tests what further calls to `NodeAddresses()` would do at that point

**Comment by danwinship** [comment]
/hold
not actually fixed (but the integration test is probably worth keeping anyway?)

**Comment by danwinship** [comment]
(This will make the unit test fail because the "cloud provider is external and nodeIP unspecified" case will now pick up the actual IP of the host it's run on.)

So do we think it's preferable to do this rather than reverting #121028?

**Comment by aojea** [comment]
> (This will make the unit test fail because the "cloud provider is external and nodeIP unspecified" case will now pick up the actual IP of the host it's run on.)
> 
> So do we think it's preferable to do this rather than reverting #121028?

/hold

good point,  I'm adding a bunch of tests so whatever we decide we have all the cases covered

**Comment by JoelSpeed** [comment]
@aojea What's the next step here, is there a need for a conversation deciding between this and the revert approach?

**Comment by aojea** [comment]
> @aojea What's the next step here, is there a need for a conversation deciding between this and the revert approach?

I need to write it down this to discuss pros and cons, it seems the filtering logic is something should not be there https://github.com/kubernetes/kubernetes/pull/125300#issuecomment-2155070827 

So I'd like anybody to agree on the next steps, there are questions like
1. do we accept filtering of addresses in a generic cloud provider code?
2. if filtering is removed, is that good enough?
3. in a kubelet that delegates in a cloud provider is ok to always perform initial discovery?
4. do we want to add the "unspecified addresses" semantics for autodiscovery?

**Comment by aojea** [comment]
/test pull-kubernetes-integration 

unrelated failure k8s.io/apiextensions-apiserver/test: integration 

I'm not in favor of reverting for two things:
1. the spirit of cloud provider external is to delegate the cloud provider to the external  controller, the flag can be used to provide an opinion.  Reverting makes Node objects that are delegated to the cloud provider to start with a IP predefined, and forces all the cloud providers to set the --node-ip flag when in theory should the other way around, people that want a specific IP should set it.
2. the --node-ip flag filtering is a bug IMHO, the justification about the CSR certificate problem only works if there are internalIPs, if there are other types of node.status.address it will fail.

I'm in favor of this because is easy to think about since it implements the same semantics we added to the --node-ip without external cloud provider and makes the perfect balance

**Review by danwinship — COMMENTED** [review]
(otherwise lgtm)

**Review by elmiko — COMMENTED** [review]
i think the rationale for this change is reasonable. the code lgtm, thanks @aojea and @danwinship , appreciate the discussion here.

**Comment by aojea** [comment]
> i think the rationale for this change is reasonable. the code lgtm, thanks @aojea and @danwinship , appreciate the discussion here.

Announced in the mailing list and created a more detailed document to reach wider audience https://groups.google.com/g/kubernetes-sig-network/c/84n7WF3T-ro/m/sE0HkK9HBQAJ and try to avoid another change

/hold cancel

File: pkg/kubelet/nodestatus/setters.go
Comment (aojea) [reply pkg/kubelet/nodestatus/setters.go:140]: updated

**Comment by aojea** [comment]
/assign @thockin @dchen1107 

this needs approval

**Comment by pacoxu** [comment]
/assign @SergeyKanzhelev 
ping @thockin @dchen1107 for approval

**Comment by pacoxu** [comment]
/priority important-longterm
or important-soon?
/triage accepted

**Review by elmiko — COMMENTED** [review]
was reviewing this again and given that we are cleaning up all the old references to in-tree cloud providers, i wonder if it makes sense to add the tests for situations where `--cloud-provider=fakeCloudTaints` ?

i am wondering if those instances where `--cloud-provider=(not "external")` shouldn't just be producing an error about in-tree providers being removed.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
